packages/block-editor/src/components/block-card/index.js

```diff
@@ -2,15 +2,20 @@
  * Internal dependencies
  */
 import BlockIcon from '../block-icon';
+import BlockTitle from '../block-title';
+import BlockDescription from '../block-description';
 
-function BlockCard( { blockType: { icon, title, description } } ) {
+function BlockCard( { clientId, blockType } ) {
+	const props = ( clientId && { clientId } ) || blockType;
 	return (
 		<div className="block-editor-block-card">
-			<BlockIcon icon={ icon } showColors />
+			<BlockIcon { ...props } showColors />
 			<div className="block-editor-block-card__content">
-				<h2 className="block-editor-block-card__title">{ title }</h2>
+				<h2 className="block-editor-block-card__title">
+					<BlockTitle { ...props } />
+				</h2>
 				<span className="block-editor-block-card__description">
-					{ description }
+					<BlockDescription { ...props } />
```

Comment:
I don't like that we're making `BlockIcon` and `BlockDescription` so polymorphic. It quickly becomes difficult to understand which level in the element tree is responsible for providing which attributes. At the level of the project, we shouldn't burden developers with choosing between passing `clientId` or explicit values.

Reply:
The initial idea I had was to unify the way how those leaf components are used. In particular, I wanted to make it easier to use in a different context where you could apply some more advanced modifications inside the component:

```js
return (
	<div className="block-editor-block-card">
		<BlockIcon clientId={ clientId } showColors />
		<div className="block-editor-block-card__content">
			<BlockTitle clientId={ clientId } as="h2" className="block-editor-block-card__title">
			<BlockDescription clientId={ clientId } className="block-editor-block-card__description" />
		</div>
	</div>
);
```

In practice, the same `BlockCard` component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to `clientId` yet.

I personally feel like developers shouldn't be concerned about deciding how for a given block to get the icon (applies to the title and description as well) and pass it to the matching component. It should accept either `clientId` or `name` and do the rest. If you think that the opposite makes more sense also because of the backward compatibility then we will have to find a way to ensure that correct data is being passed in all places. The biggest challenge is the ambiguity we introduce by having `getBlockType` that returns the original block definition and the new method that combines that with the active block variation that can be only achieved by providing the `clientId`.

Reply:
>At the level of the project, we shouldn't burden developers with choosing between passing clientId or explicit values.

I personally don't think it's will be a burden to devs, quite the opposite.. This way we have a consistent way(using the three components) of displaying block's info and could even be more easy to use with the fallback of getting the `selectedBlockClientId`. For example if we need to show info on a selected block it could be just `<BlockDescription />`.

Reply:
I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */
    <BlockIcon showColors />
    <BlockTitle as="h2" />
```

Reply:
(edit: I accidentally sent this comment before finishing it)

I think you two make good arguments, leaving me undecided. I have a follow-up question, but let me know if it's too farfetched:

What if `BlockIcon`, `BlockTitle`, `BlockDescription` are made to be contextual:

```jsx
<BlockEditContextProvider value={ ... }> /* implicit ancestor */

        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
```

... with the provision that `BlockIcon` accepts an explicit `blockName` prop as an override:

```js
BlockIcon = ( { blockName } ) => {
    const { clientId } = useBlockEditContext();
    const icon = blockName ?
        grabIconFromBlockType( blockName ) :
        grabIconFromBlockInstance( clientId );
    /* ... */
```

That way, `BlockCard` might choose to provide its own context in order to address:

> In practice, the same BlockCard component is used not only in the block sidebar but also in the block preview in the inserter where you don't have access to clientId yet.

```js
BlockCard = ( props ) => {
    const element = (
        <div className="block-editor-block-card">
            <BlockIcon showColors />
            <BlockTitle as="h2" />
        </div>
    );

    if ( empty( props ) ) {
        return element;
    }

    return (
        <BlockEditContextProvider value={ props }>
             { element }
        </BlockEditContextProvider>       
    );
}
```

Reply:
I wouldn't mind if we would take advantage of `useBlockEditContext` here, we do it in other places in `@wordpress/block-editor` since it holds both the block's `name` and `clientId`.

https://github.com/WordPress/gutenberg/blob/18349c49bab0d9a227e2608a66420f7a9f6e70ac/packages/block-editor/src/components/block-edit/context.js#L12-L16

Reply:
Me neither! I can try these changes, along with some of the other comments and take it from there.. :)

Reply:
Me neither! I can try these changes, along with some of the other comments and take it from there.. :)